lwrunner/awsrunner.go

```diff
@@ -116,6 +122,180 @@ func (run AWSRunner) SendPublicKey(pubBytes []byte) error {
+		return nil
+	} else { // associate our own instance profile
+		associateInput := &ec2.AssociateIamInstanceProfileInput{
+			IamInstanceProfile: &ec2types.IamInstanceProfileSpecification{
+				Arn: instanceProfile.Arn,
+			},
+			InstanceId: aws.String(run.InstanceID),
+		}
+		_, err = c.AssociateIamInstanceProfile(context.Background(), associateInput)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+}
+
+// isCorrectInstanceProfileAlreadyAssociated takes a list of instance profile associations
+// and checks if there is an instance profile associated and if this instance
+// profile has the correct policy for SSM access. Returns `true, nil` if so. Returns
+// `false, nil` if there is no instance profile associated. Returns `false, <error>` if
+// there is an incorrect instance profile associated, or if there was an error in
+// executing this function.
+func (run AWSRunner) isCorrectInstanceProfileAlreadyAssociated(cfg aws.Config, associations []ec2types.IamInstanceProfileAssociation) (bool, error) {
+	if len(associations) <= 0 { // no instance profile associated
+		return false, nil
+	}
+	instanceProfileName := strings.Split(*associations[0].IamInstanceProfile.Arn, "instance-profile/")[1]
+
+	c := iam.New(iam.Options{
+		Credentials: cfg.Credentials,
+		Region:      cfg.Region,
+	})
+
+	getInstanceProfileInput := &iam.GetInstanceProfileInput{
+		InstanceProfileName: aws.String(instanceProfileName),
+	}
+	getInstanceProfileOutput, err := c.GetInstanceProfile(context.Background(), getInstanceProfileInput)
+	if err != nil {
+		return false, err
+	}
+
+	// Check to see if the instance profile associated with the runner has the correct policy
+
+	if len(getInstanceProfileOutput.InstanceProfile.Roles) <= 0 { // can only have max one role
+		return false, fmt.Errorf(
+			"runner %v already has an instance profile (%v) attached, does not have a role",
+			run,
+			getInstanceProfileOutput.InstanceProfile,
+		)
+	}
+
+	// Check which policies are associated with this instance profile's role
+	listAttachedRolePoliciesInput := iam.ListAttachedRolePoliciesInput{
+		RoleName: getInstanceProfileOutput.InstanceProfile.Roles[0].RoleName,
+	}
+	listAttachedRolePoliciesOutput, err := c.ListAttachedRolePolicies(context.Background(), &listAttachedRolePoliciesInput)
+	if err != nil {
+		return false, err
+	}
+
+	for _, policy := range listAttachedRolePoliciesOutput.AttachedPolicies {
+		if *policy.PolicyArn == SSMInstancePolicy {
+			return true, nil // everything is configured correctly, we can return now
+		}
+	}
+
+	// The runner has an instance profile attached, the instance profile has a role,
+	// and the role does not have the policy we need for SSM. We can't install on
+	// this instance, return an error
+	return false, fmt.Errorf(
+		"runner %v already has an instance profile (%v) attached, does not have policy %s",
+		run,
+		getInstanceProfileOutput.InstanceProfile,
+		SSMInstancePolicy,
+	)
```

Comment:
fwiw in future work we could still attach policies to existing roles: https://docs.aws.amazon.com/sdk-for-go/api/service/iam/#AttachRolePolicyInput . I _do_ think it should be future work, checking all that and cleaning it up is a bunch of busywork, and what we already have here is great and useful.

Reply:
Totally agree, I filed a ticket for this, https://lacework.atlassian.net/browse/RAIN-47063